### PR TITLE
OSGi support

### PR DESCRIPTION
**Pull request by gnodet**
Support for OSGi.
Requires https://github.com/wildfly/wildfly-client-config/pull/5.

### CODE REVIEW

**Comment by undertow-pull-request** [comment]
Can one of the admins verify this patch?

**Comment by ctomc** [comment]
this is ok to test

**Comment by undertow-pull-request** [comment]
**Linux** [Build 2207](https://ci.wildfly.org/viewLog.html?buildId=44793&buildTypeId=UT_MASTER_LINUX) outcome was **FAILURE** using a merge of 2d62486ed414049911b874be96e011e95063ebde
Summary: Exit code 1 (new) Build time: 00:00:11

**Comment by undertow-pull-request** [comment]
**Windows** [Build 1944](https://ci.wildfly.org/viewLog.html?buildId=44794&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of 2d62486ed414049911b874be96e011e95063ebde
Summary: Exit code 1 Build time: 00:00:19

File: pom.xml [hunk 1]
```diff
@@ -70,14 +70,17 @@
...
+        <version.org.jboss.logging>3.3.0.Final</version.org.jboss.logging>
         <version.org.jboss.logging.processor>2.0.0.Final</version.org.jboss.logging.processor>
         <version.org.jboss.logmanager>2.0.0.Final</version.org.jboss.logmanager>
         <version.org.jboss.spec.javax.servlet.jboss-servlet-api_4.0_spec>1.0.0.Alpha1</version.org.jboss.spec.javax.servlet.jboss-servlet-api_4.0_spec>
         <version.org.jboss.spec.javax.annotation.jboss-annotations-api_1.2_spec>1.0.0.Final</version.org.jboss.spec.javax.annotation.jboss-annotations-api_1.2_spec>
         <version.org.jboss.spec.javax.servlet.jsp>1.0.0.Final</version.org.jboss.spec.javax.servlet.jsp>
         <version.org.jboss.spec.javax.websockets>1.1.0.Final</version.org.jboss.spec.javax.websockets>
-        <version.xnio>3.3.6.Final</version.xnio>
+        <version.xnio>3.5.0.Beta1</version.xnio>
+        <version.org.wildfly.common>1.2.0.Beta6-SNAPSHOT</version.org.wildfly.common>
```
Comment (stuartwdouglas) [inline pom.xml]: Why have you added a dependency on these Wildfly components?

File: servlet/pom.xml [hunk 2]
```diff
@@ -65,6 +65,13 @@
             <groupId>org.jboss.spec.javax.annotation</groupId>
             <artifactId>jboss-annotations-api_1.2_spec</artifactId>
         </dependency>
+
+        <dependency>
+            <groupId>org.osgi</groupId>
```
Comment (stuartwdouglas) [inline servlet/pom.xml]: This should be an optional dependency.

File: servlet/src/main/java/io/undertow/servlet/core/ServletExtensionHolder.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+package io.undertow.servlet.core;
+
+import java.util.List;
+import java.util.concurrent.CopyOnWriteArrayList;
+
+import io.undertow.servlet.ServletExtension;
+
+public class ServletExtensionHolder {
+
+    private static List<ServletExtension> extensions = new CopyOnWriteArrayList<>();
```
Comment (stuartwdouglas) [inline servlet/src/main/java/io/undertow/servlet/core/ServletExtensionHolder.java]: Why have you added this? Undertow does not really use any static config, all config is passed in at deployment time.

File: servlet/src/main/java/io/undertow/servlet/osgi/Activator.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+import io.undertow.servlet.ServletExtension;
+import io.undertow.servlet.core.ServletExtensionHolder;
+import org.osgi.framework.BundleActivator;
+import org.osgi.framework.BundleContext;
+import org.osgi.framework.ServiceReference;
+import org.osgi.util.tracker.ServiceTracker;
+import org.osgi.util.tracker.ServiceTrackerCustomizer;
+
+public class Activator implements BundleActivator, ServiceTrackerCustomizer<ServletExtension, ServletExtension> {
```
Comment (stuartwdouglas) [inline servlet/src/main/java/io/undertow/servlet/osgi/Activator.java]: What does this do?

**Comment by undertow-pull-request** [comment]
**Linux** [Build 2262](https://ci.wildfly.org/viewLog.html?buildId=53429&buildTypeId=UT_MASTER_LINUX) outcome was **FAILURE** using a merge of 640afe213ae3c8ac425580bae29a0dffc699a1c6
Summary: Exit code 1 Build time: 00:00:12

**Comment by undertow-pull-request** [comment]
**Windows** [Build 1999](https://ci.wildfly.org/viewLog.html?buildId=53430&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of 640afe213ae3c8ac425580bae29a0dffc699a1c6
Summary: Exit code 1 Build time: 00:01:05

File: servlet/src/main/java/io/undertow/servlet/core/ServletExtensionHolder.java (hunk 3 above)
Comment (gnodet) [inline servlet/src/main/java/io/undertow/servlet/core/ServletExtensionHolder.java]: This is for the websocket Bootstrap class to be loaded in OSGi.

File: servlet/src/main/java/io/undertow/servlet/osgi/Activator.java (hunk 4 above)
Comment (gnodet) [inline servlet/src/main/java/io/undertow/servlet/osgi/Activator.java]: I've added more docs in the latest commits.

File: pom.xml (hunk 1 above)
Comment (gnodet) [inline pom.xml]: Because older versions do not have OSGi metadata.

File: servlet/pom.xml (hunk 2 above)
Comment (gnodet) [inline servlet/pom.xml]: It's at scope provided, which means it does not show as a transitive dependency.  I can make it provided + optional if you prefer.

**Comment by undertow-pull-request** [comment]
**Linux** [Build 2267](https://ci.wildfly.org/viewLog.html?buildId=53578&buildTypeId=UT_MASTER_LINUX) outcome was **FAILURE** using a merge of 98f8655ed3cff542e3420b38fca276d10654fcae
Summary: Exit code 1 Build time: 00:00:31

**Comment by undertow-pull-request** [comment]
**Windows** [Build 2004](https://ci.wildfly.org/viewLog.html?buildId=53579&buildTypeId=UNDERTOW_MasterWindows) outcome was **FAILURE** using a merge of 98f8655ed3cff542e3420b38fca276d10654fcae
Summary: Exit code 1 Build time: 00:00:32

**Comment by gnodet** [comment]
Merged with https://github.com/undertow-io/undertow/commit/cca6ca9706a62b268a5e6b92bbc71cb5ad23df7c
Thx a ton !

**Pull request closed without merging**